packages/next-server/lib/head.tsx

```diff
@@ -137,32 +137,62 @@ function reduceComponents(
     })
 }
 
-const Effect = withSideEffect()
-
 /**
  * This component injects elements to `<head>` of your page.
  * To avoid duplicated `tags` in `<head>` you can use the `key` property, which will make sure every tag is only rendered once.
  */
-function Head({ children }: { children: React.ReactNode }) {
-  return (
-    <AmpStateContext.Consumer>
-      {ampState => (
-        <HeadManagerContext.Consumer>
-          {updateHead => (
-            <Effect
-              reduceComponentsToState={reduceComponents}
-              handleStateChange={updateHead}
-              inAmpMode={isInAmpMode(ampState)}
-            >
-              {children}
-            </Effect>
-          )}
-        </HeadManagerContext.Consumer>
-      )}
-    </AmpStateContext.Consumer>
-  )
+function Head({ children }: { children: JSX.Element }) {
+  const ampState = React.useContext(AmpStateContext)
+  const updateHead = React.useContext(HeadManagerContext)
+  const instanceRef = React.useRef(children)
+
+  // Update the instanceRef every render
+  instanceRef.current = children
+
+  const inAmpMode = isInAmpMode(ampState)
+
+  const emitUpdate = () => {
+    headState = reduceHeadInstances([...headInstanceRefs], {
```

Comment:
This would be better done by making `reduceHeadInstances` accept `Iterable` and passing it through `Array.from` before reducing.